corehq/apps/locations/bulk_management.py

```diff
@@ -0,0 +1,131 @@
+"""
+Bulk rearrange locations.
+
+This includes support for changing location types, changing locations' parents,
+deleting things, and so on.  See the spec doc for specifics:
+https://docs.google.com/document/d/1gZFPP8yXjPazaJDP9EmFORi88R-jSytH6TTgMxTGQSk/
+"""
+from collections import namedtuple, Counter, defaultdict
+
+from dimagi.utils.decorators.memoized import memoized
+
+from .tree_utils import BadParentError, CycleError, assert_no_cycles
+
+LocationTypeStub = namedtuple(
+    "LocationTypeStub",
+    "name code parent_code shares_cases view_descendants"
+)
+
+LocationStub = namedtuple(
+    "LocationStub",
+    "name site_code location_type parent_code location_id external_id latitude longitude"
+)
+
+
+class LocationTreeValidator(object):
+    def __init__(self, location_types, locations):
+        self.location_types = location_types
+        self.locations = locations
+        self.types_by_code = {lt.code: lt for lt in location_types}
+        self.locations_by_code = {l.site_code: l for l in locations}
+
+    @property
+    @memoized
+    def errors(self):
+        # We want to find as many errors as possible up front, but some high
+        # level errors make it unrealistic to keep validating
+        basic_errors = (self._check_unique_type_codes() +
+                        self._check_unique_location_codes())
+        if basic_errors:
+            # it doesn't make sense to try to validate a tree when you can't
+            # uniquely determine the relationships
+            return basic_errors
+
+        # Make sure the location types make sense
+        type_errors = self._validate_location_types()
+        if type_errors:
+            return type_errors
+
+        # Check each location's position in the tree
+        errors = filter(None, [
```

Comment:
All descendants of an invalid location are invalid too, how does that get handled?